docs/architecture/adr-022-custom-panic-handling.md

```diff
@@ -0,0 +1,155 @@
+
+It will also make `OutOfGas` case and `default` case one of the middlewares.
+
+## Decision
+
+### Design
+
+#### Overview
+
+Instead of hardcoding custom error handling into BaseApp we suggest using set of middlewares which can be customized
+externally and will allow developers use as many custom error handlers as they want. Implementation with tests
+can be found [here](https://github.com/cosmos/cosmos-sdk/pull/6053).
+
+#### Implementation details
+
+##### RecoveryHandler
+
+We add a `recover()` object handler type:
+
+```go
+type RecoveryHandler func(recoveryObj interface{}) error
+```
+
+This handler receives an object, processes it and returns contextually enriched `error` or `nil`
+if object is not a handle's target type.
+
+An example:
+
+```go
+func(recoveryObj interface{}) error {
+    err, ok := recoveryObj.(error)
+    if !ok { return nil }
+    
+    if someSpecificError.Is(err) {
+        panic(customPanicMsg)
+    } else {
+        return nil
+    }
+}
+```
+
+This example breaks the application execution, but it also might enrich the error's context like the `OutOfGas` handler.
+
+##### RecoveryMiddleware
+
+We also add a middleware type:
+
+```go
+type recoveryMiddleware func(recoveryObj interface{}) (recoveryMiddleware, error)
+```
+
+Function receives a `recover()` object and returns (`next middleware`, `nil`) if object wasn't handled (not a target type)
+or (`nil`, `error`) if input object was handled and other middlewares in the chain should not be executed.
+
+`OutOfGas` middleware example:
+```go
+func newRecoveryMiddleware(handler RecoveryHandler, next recoveryMiddleware) recoveryMiddleware {
+    return func(recoveryObj interface{}) (recoveryMiddleware, error) {
+        if err := handler(recoveryObj); err != nil {
+            return nil, err
+        }
+        return next, nil
+    }
+}
+
+func newOutOfGasRecoveryMiddleware(gasWanted uint64, ctx sdk.Context, next recoveryMiddleware) recoveryMiddleware {
+    handler := func(recoveryObj interface{}) error {
+        err, ok := recoveryObj.(sdk.ErrorOutOfGas)
+        if !ok { return nil }
+
+        return sdkerrors.Wrap(
+            sdkerrors.ErrOutOfGas, fmt.Sprintf(
+                "out of gas in location: %v; gasWanted: %d, gasUsed: %d", err.Descriptor, gasWanted, ctx.GasMeter().GasConsumed(),
+            ),
+        )
+    }
+    
+    return newRecoveryMiddleware(handler, next)
+}
+```
+
+##### Recovery processing
+
+Basic chain of middlewares processing would look like:
+
+```go
+func processRecovery(recoveryObj interface{}, middleware recoveryMiddleware) error {
+	if middleware == nil { return nil }
+
+	next, err := middleware(recoveryObj)
+	if err != nil { return err }
+
+	if next != nil { return processRecovery(recoveryObj, next) }
+
+	return nil
+}
+```
+
+That way we can create a middleware chain which is executed from left to right, the rightmost middleware is a
+`default` handler which must return an `error`.
+
+##### Baseapp changes
```

Comment:
I assume this will be set through a baseapp option on the application constructor? I'd add that section too.